plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/type/TestTypeInfoUtils.java

```diff
@@ -0,0 +1,207 @@
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: 'type' expected at the position 10 of 'TIMESTAMP WITH TIME ZONE' but 'WITH' is found.");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("ARRAY"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: < expected at the end of 'ARRAY'");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("STRUCT"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: < expected at the end of 'STRUCT'");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("STRUCT<INT64, STRING>"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: ' ' expected at the position 12 of 'STRUCT<INT64, STRING>' but ',' is found.");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("STRUCT<# INT64>"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: 'name' expected at the position 7 of 'STRUCT<# INT64>' but '#' is found.");
+
+        // leading whitespace is not permitted, but trailing is truncated
+        assertThatThrownBy(() -> getTypeInfoFromTypeString(" JSON"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: 'type' expected at the position 0 of ' JSON' but ' ' is found.");
+        assertTypeInfo("JSON ", "JSON");
+
+        // invalid type parameters should not cause out of bounds errors
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("STRING("))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("type parameters expected for type string STRING(");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("STRING()"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("type parameters expected for type string STRING()");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("STRING(1, 2)"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Type string only takes zero or one parameter, but 2 is seen");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("NUMERIC(1, 2, 3)"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Type decimal only takes two parameters, but 3 is seen");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("ARRAY<"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: type expected at the end of 'ARRAY<'");
+        assertThatThrownBy(() -> getTypeInfoFromTypeString("ARRAY<>"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Error: 'type' expected at the position 6 of 'ARRAY<>' but '>' is found.");
+
+        // handle special characters
```

Comment:
I don't think BigQuery returns these values on `data_type` columns. These unit test should be rewritten as field names on struct types. Also, we should test with actual BigQuery service as the value may contain additional quotes or something. 

It would be nice to add more test cases. Please refer to `BaseConnectorTest#testColumnNameTestData`. 

Also, I found the new implementation has a bug when the field name contains a space.

Reply:
BigQuery allows [flexible column names](https://cloud.google.com/bigquery/docs/schemas#column_names), and when used in struct fields, they're not quoted or escaped.

I was able to create a table with a struct field with another type embedded in the name (`type inside array<int64> name`), and we're reading it back like this:
```
STRUCT<type inside array<int64> name STRING>
```

I'm not sure how to continue with this. I don't think we can correctly handle this in the parser. Would it make sense to make the usage of `streamRelationColumns` opt-int?

Reply:
We can also try a hybrid solution - ignore `STRUCT` in the parser and get these field types from the BQ API.

Reply:
All done, please review again since the changes are significant. I used a supplier to avoid calling `client.getTable()` if there are no struct types used. I also noticed we might have been returning incorrect schema names (remote vs local).

Funny thing is that Trino also produces unquoted field names in row types in `information_schema.columns` :-)